### ISSUE TITLE
no-asm sanitizer 04-test_encoder_decoder test is *really* slow

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#13495]
This test seems to generate a large number of keys before it starts which takes a *long* time (particularly with no-asm/sanitizer builds). Can we somehow avoid the key gen step?

**Comment by levitte** [comment]
Sure, we can generate the keys once and have them on file. It just seemed a lite weird to decode stuff just to have them encoded and decoded again.
I was thinking for a bit that since the first step of each test case is to encode the key, we could make that a one-time thing... except that we do use the original key to compare the decoding result against, so that would diminish the test for the lack of original data... 

That's how far I have gotten in thinking about this. 

I do question the need for speedier tests, though (there's a definite trend going on right now, at least in the minds of some). In the choice between speed and reliability, I wouldn't think twice, the latter wins.

**Comment by mattcaswell** [comment]
>Sure, we can generate the keys once and have them on file. It just seemed a lite weird to decode stuff just to have them encoded and decoded again.

I can see that - but if it doesn't break the purpose of the test, then this would seem like the preferred way forward. An alternative is to construct the keys "manually" in the code itself.

> I do question the need for speedier tests

I am strongly in favour of faster tests. We absolutely *must* be careful not to kill the performance of the test suite unnecessarily. The intention I believe is to continue invest in our test suite and expand the testing considerably as we move forwards. The slower the tests are the more this will become a huge problem for us - both in terms of the resources required to run our tests as well as the time taken. I don't know about you but I run the test suite many times a day. It doesn't take much to multiply that up and equate to large amounts of time lost.

>  In the choice between speed and reliability, I wouldn't think twice, the latter wins.

I don't see it as an either/or situation. We *must* have test coverage and  it *must* be as performant as possible.

**Comment by kroeckx** [comment]
We now have this concept of extended tests in CI, but a default "make test" runs all the tests. And I actually want the default to have all the tests.

Are there some tests that we think take long to run and in most cases are not useful to run? Should we then have an easy way to exclude those tests?

**Comment by kaduk** [comment]
My normal build process uses `TESTS=-test_fuzz`, but skips the external tests because they are not present/configured.  The CI extended tests include those external tests as one of (IIUC) the main differences from the non-extended tests.

**Comment by DDvO** [comment]
On my machine `time make test TESTS=test_encoder_decoder` varies between 10 and 20 seconds,
and when the sanitizers are enabled this time doubles: between 20 and 44 seconds. 
Which is a lot, in particular since it causes 100% load (for one CPU core).

**Comment by levitte** [comment]
Something I observed when looking at the github CI builds is that `no-asm` is really slow. I never combine that with sanitizers on my machine, which might be a reason why I never experienced the excruciating slowness...

**Comment by t8m** [comment]
Yeah, it would make sense to not use `no-asm` with the sanitizers build.

**Comment by kroeckx** [comment]
> Yeah, it would make sense to not use `no-asm` with the sanitizers build.

A no-asm sanitizer build is useful, because the C code then
actually gets build and tested. If you use the assembler, the
assembler doesn't get sanitized.

**Comment by mattcaswell** [comment]
> A no-asm sanitizer build is useful, because the C code then
actually gets build and tested. If you use the assembler, the
assembler doesn't get sanitized.

But is that something we need to run on every PR? We could have a run-checker job that does this.

**Comment by levitte** [comment]
> But is that something we need to run on every PR? We could have a run-checker job that does this.

Yes!

**Comment by mattcaswell** [comment]
> On my machine time make test TESTS=test_encoder_decoder varies between 10 and 20 seconds,
and when the sanitizers are enabled this time doubles: between 20 and 44 seconds.
Which is a lot, in particular since it causes 100% load (for one CPU core).

I ran some of my own tests on my laptop. I just did one run in each configuration so there may be some variability but it gives some kind of picture. I built with the following configurations:

````
./config -d --strict-warnings
./config -d --strict-warnings no-asm
./config -d --strict-warnings enable-asan enable-ubsan
./config -d --strict-warnings enable-ubsan enable-asan no-asm
````
I then ran just the TEST_encoder_decoder test and timed the result with `time`. The results were:

No options: 0m5.308s
no-asm: 7m45.533s
Sanitizers: 0m36.559s
Sanitizers plus no-asm: 45m54.454s

So approximately speaking it looks to me like the sanitizers add an overhead of approx. x7, while no-asm adds approximately x88. The two together gives a slow down of approx. x500.

45 mins to run one test seems unreasonably long!! I think it is primarily due to the large number of keys that this test generates before it starts. I think we should:

1) Move the no-asm plus sanitizers CI jobs to be run-checker only
2) Refactor the test to not generate large numbers of keys up front.

**Comment by DDvO** [comment]
> Refactor the test to not generate large numbers of keys up front.

It turns out that in `endecode_test.c` it's just 14 keys, and only the first two (DH and DHX) are very slow, in particular the DH one.
For a significant speedup it should be sufficient to use for them pre-established parameters (or short key sizes).

**Comment by DDvO** [comment]
My suspicion was right, see #13552.

**Comment by kroeckx** [comment]
Note that various of the travis runs are failing because they timeout after 50 minutes, and all except the first job seems to be fixed by #13552.

**Comment by DDvO** [comment]
Re-opening this issue because the following is not yet done:

Move no-asm sanitizer CI builds to a run-checker job.

**Comment by levitte** [comment]
@DDvO, thanks for the deeper analysis.  I had a first look at #13552 yesterday or so, and am very pleased with the solution...  I was afraid we'd had to sabotage the test to get it running in short enough time...

**Comment by DDvO** [comment]
@levitte, pleased to hear.
Will you take care of moving the no-asm sanitizer builds to run-checker?

**Comment by mattcaswell** [comment]
> Will you take care of moving the no-asm sanitizer builds to run-checker?

https://github.com/openssl/tools/pull/77

**Comment by DDvO** [comment]
Closing again since meanwhile also the remaining issue has been solved by openssl/tools#77.
